In [1]:
import numpy as np
import pandas as pd


In [2]:
url = "https://raw.githubusercontent.com/dineshpiyasamara/LaptopPricePredictor/master/model%20building/laptop_price.csv"
data = pd.read_csv(url,encoding='latin-1')

In [3]:
data.head()

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   
3          4   Apple  MacBook Pro  Ultrabook    15.4   
4          5   Apple  MacBook Pro  Ultrabook    13.3   

                     ScreenResolution                         Cpu   Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

                            Gpu  OpSys  Weight  Price_euros  
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69  
1        Intel HD Graphics 6000  macOS  1.34kg       898.94  
2         Intel HD Graphics 620  No OS  1.86kg       575.00  
3            AMD Radeon Pro 455  macOS  1.83kg      2537.45  
4  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60

In [4]:
data.shape

(1303, 12)

In [6]:
data.isnull().sum()

laptop_ID           0
Company             0
Product             0
TypeName            0
Inches              0
ScreenResolution    0
Cpu                 0
Ram                 0
Gpu                 0
OpSys               0
Weight              0
Price_euros         0
dtype: int64

There are no null values

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         1303 non-null   int64  
 1   Company           1303 non-null   object 
 2   Product           1303 non-null   object 
 3   TypeName          1303 non-null   object 
 4   Inches            1303 non-null   float64
 5   ScreenResolution  1303 non-null   object 
 6   Cpu               1303 non-null   object 
 7   Ram               1303 non-null   object 
 8   Gpu               1303 non-null   object 
 9   OpSys             1303 non-null   object 
 10  Weight            1303 non-null   object 
 11  Price_euros       1303 non-null   float64
dtypes: float64(2), int64(1), object(9)
memory usage: 122.3+ KB


In [9]:
data.head(2)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   

                     ScreenResolution                   Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600  Intel Core i5 2.3GHz  8GB   
1                            1440x900  Intel Core i5 1.8GHz  8GB   

                            Gpu  OpSys  Weight  Price_euros  
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69  
1        Intel HD Graphics 6000  macOS  1.34kg       898.94

Laptop ID have to be removed, it has no effect on the data

In [10]:
data['Ram'] = data['Ram'].str.replace('GB','').astype('int32')
data['Weight'] = data['Weight'].str.replace('kg','').astype('float32')

In [11]:
data.head(2)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   

                     ScreenResolution                   Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600  Intel Core i5 2.3GHz    8   
1                            1440x900  Intel Core i5 1.8GHz    8   

                            Gpu  OpSys  Weight  Price_euros  
0  Intel Iris Plus Graphics 640  macOS    1.37      1339.69  
1        Intel HD Graphics 6000  macOS    1.34       898.94

In [12]:
data.corr()['Price_euros']

<ipython-input-12-f08e50fc67e8>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  data.corr()['Price_euros']


laptop_ID      0.067830
Inches         0.068197
Ram            0.743007
Weight         0.210370
Price_euros    1.000000
Name: Price_euros, dtype: float64

Clearly LaptopID and Inches have no correlation
can only find correlation value on numerical values

## Company Column

In [13]:
data['Company'].value_counts()

Dell         297
Lenovo       297
HP           274
Asus         158
Acer         103
MSI           54
Toshiba       48
Apple         21
Samsung        9
Razer          7
Mediacom       7
Microsoft      6
Xiaomi         4
Vero           4
Chuwi          3
Google         3
Fujitsu        3
LG             3
Huawei         2
Name: Company, dtype: int64

In [14]:
def add_company(inpt):
    if inpt == 'Samsung' or inpt == 'Razer' or inpt == 'Mediacom' or inpt == 'Microsoft'or inpt == 'Xiaomi'or inpt == 'Vero'or inpt == 'Chuwi'or inpt == 'Google'or inpt == 'Fujitsu'or inpt == 'LG'or inpt == 'Huawei':
        return 'Other'
    else:
        return inpt

In [15]:
data['Company'] = data['Company'].apply(add_company)

In [16]:
data['Company'].value_counts()

Dell       297
Lenovo     297
HP         274
Asus       158
Acer       103
MSI         54
Other       51
Toshiba     48
Apple       21
Name: Company, dtype: int64

## Product Column

In [19]:
data['Product'].value_counts()

XPS 13                                     30
Inspiron 3567                              29
250 G6                                     21
Legion Y520-15IKBN                         19
Vostro 3568                                19
                                           ..
15-bw007nv (A10-9620P/6GB/128GB/Radeon      1
SmartBook 130                               1
A541NA-GO342 (N3350/4GB/500GB/Linux)        1
17-X047na (i3-6006U/8GB/1TB/W10)            1
V330-15IKB (i5-8250U/4GB/500GB/FHD/W10)     1
Name: Product, Length: 618, dtype: int64

There are 618 Different (catergories) values under product ==> Have no effect on the price of the laptop

## TypeName Column

In [20]:
data['TypeName'].value_counts()

Notebook              727
Gaming                205
Ultrabook             196
2 in 1 Convertible    121
Workstation            29
Netbook                25
Name: TypeName, dtype: int64

## ScreenResolution Column

In [21]:
data['ScreenResolution'].value_counts()

Full HD 1920x1080                                507
1366x768                                         281
IPS Panel Full HD 1920x1080                      230
IPS Panel Full HD / Touchscreen 1920x1080         53
Full HD / Touchscreen 1920x1080                   47
1600x900                                          23
Touchscreen 1366x768                              16
Quad HD+ / Touchscreen 3200x1800                  15
IPS Panel 4K Ultra HD 3840x2160                   12
IPS Panel 4K Ultra HD / Touchscreen 3840x2160     11
4K Ultra HD / Touchscreen 3840x2160               10
4K Ultra HD 3840x2160                              7
Touchscreen 2560x1440                              7
IPS Panel 1366x768                                 7
IPS Panel Quad HD+ / Touchscreen 3200x1800         6
IPS Panel Retina Display 2560x1600                 6
IPS Panel Retina Display 2304x1440                 6
Touchscreen 2256x1504                              6
IPS Panel Touchscreen 2560x1440               

In [22]:
data['Touchscreen'] = data['ScreenResolution'].apply(lambda x:1 if 'Touchscreen' in x else 0)
data['Ips'] = data['ScreenResolution'].apply(lambda x:1 if 'IPS' in x else 0)

In [23]:
data.head(2)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   

                     ScreenResolution                   Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600  Intel Core i5 2.3GHz    8   
1                            1440x900  Intel Core i5 1.8GHz    8   

                            Gpu  OpSys  Weight  Price_euros  Touchscreen  Ips  
0  Intel Iris Plus Graphics 640  macOS    1.37      1339.69            0    1  
1        Intel HD Graphics 6000  macOS    1.34       898.94            0    0

## Cpu Column

In [24]:
data['Cpu'].value_counts()

Intel Core i5 7200U 2.5GHz       190
Intel Core i7 7700HQ 2.8GHz      146
Intel Core i7 7500U 2.7GHz       134
Intel Core i7 8550U 1.8GHz        73
Intel Core i5 8250U 1.6GHz        72
                                ... 
Intel Core M M3-6Y30 0.9GHz        1
AMD A9-Series 9420 2.9GHz          1
Intel Core i3 6006U 2.2GHz         1
AMD A6-Series 7310 2GHz            1
Intel Xeon E3-1535M v6 3.1GHz      1
Name: Cpu, Length: 118, dtype: int64

In [25]:
data['cpu_name'] = data['Cpu'].apply(lambda x:" ".join(x.split()[0:3]))

In [26]:
data['cpu_name'].value_counts()

Intel Core i7               527
Intel Core i5               423
Intel Core i3               136
Intel Celeron Dual           80
Intel Pentium Quad           27
Intel Core M                 19
AMD A9-Series 9420           12
Intel Celeron Quad            8
AMD A6-Series 9220            8
AMD A12-Series 9720P          7
Intel Atom x5-Z8350           5
AMD A8-Series 7410            4
Intel Atom x5-Z8550           4
Intel Pentium Dual            3
AMD A9-Series 9410            3
AMD Ryzen 1700                3
AMD A9-Series A9-9420         2
AMD A10-Series 9620P          2
Intel Atom X5-Z8350           2
AMD E-Series E2-9000e         2
Intel Xeon E3-1535M           2
Intel Xeon E3-1505M           2
AMD E-Series 7110             2
AMD A10-Series 9600P          2
AMD A6-Series A6-9220         2
AMD A10-Series A10-9620P      2
AMD Ryzen 1600                1
Intel Atom x5-Z8300           1
AMD E-Series E2-6110          1
AMD FX 9830P                  1
AMD E-Series E2-9000          1
AMD A6-S

In [27]:
def set_processor(name):
    if name == 'Intel Core i7' or name == 'Intel Core i5' or name == 'Intel Core i3':
        return name
    else:
        if name.split()[0] == 'AMD':
            return 'AMD'
        else:
            return 'Other'

In [28]:
data['cpu_name'] = data['cpu_name'].apply(set_processor)

In [29]:
data['cpu_name'].value_counts()

Intel Core i7    527
Intel Core i5    423
Other            155
Intel Core i3    136
AMD               62
Name: cpu_name, dtype: int64

## Ram Column

In [30]:
data['Ram'].value_counts()

8     619
4     375
16    200
6      41
12     25
2      22
32     17
24      3
64      1
Name: Ram, dtype: int64

## Gpu Column

In [31]:
data['Gpu'].value_counts()

Intel HD Graphics 620      281
Intel HD Graphics 520      185
Intel UHD Graphics 620      68
Nvidia GeForce GTX 1050     66
Nvidia GeForce GTX 1060     48
                          ... 
AMD Radeon R5 520            1
AMD Radeon R7                1
Intel HD Graphics 540        1
AMD Radeon 540               1
ARM Mali T860 MP4            1
Name: Gpu, Length: 110, dtype: int64

In [32]:
data['gpu_name'] = data['Gpu'].apply(lambda x:" ".join(x.split()[0:1]))

In [33]:
data['gpu_name'].value_counts()

Intel     722
Nvidia    400
AMD       180
ARM         1
Name: gpu_name, dtype: int64

In [34]:
data.shape

(1303, 16)

In [36]:
data = data[data['gpu_name'] != 'ARM']

Removed the ARM

In [37]:
data.shape

(1302, 16)

In [35]:
data.head(2)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   

                     ScreenResolution                   Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600  Intel Core i5 2.3GHz    8   
1                            1440x900  Intel Core i5 1.8GHz    8   

                            Gpu  OpSys  Weight  Price_euros  Touchscreen  Ips  \
0  Intel Iris Plus Graphics 640  macOS    1.37      1339.69            0    1   
1        Intel HD Graphics 6000  macOS    1.34       898.94            0    0   

        cpu_name gpu_name  
0  Intel Core i5    Intel  
1  Intel Core i5    Intel

## OpSys Column

In [38]:
data['OpSys'].value_counts()

Windows 10      1072
No OS             66
Linux             62
Windows 7         45
Chrome OS         26
macOS             13
Mac OS X           8
Windows 10 S       8
Android            2
Name: OpSys, dtype: int64

In [39]:
def set_os(inpt):
    if inpt == 'Windows 10' or inpt == 'Windows 7' or inpt == 'Windows 10 S':
        return 'Windows'
    elif inpt == 'macOS' or inpt == 'Mac OS X':
        return 'Mac'
    elif inpt == 'Linux':
        return inpt
    else:
        return 'Other'

In [41]:
data['OpSys'] = data['OpSys'].apply(set_os)

<ipython-input-41-66a3f027e424>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['OpSys'] = data['OpSys'].apply(set_os)


In [42]:
data['OpSys'].value_counts()

Other    1240
Linux      62
Name: OpSys, dtype: int64

In [43]:
data.head(2)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   

                     ScreenResolution                   Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600  Intel Core i5 2.3GHz    8   
1                            1440x900  Intel Core i5 1.8GHz    8   

                            Gpu  OpSys  Weight  Price_euros  Touchscreen  Ips  \
0  Intel Iris Plus Graphics 640  Other    1.37      1339.69            0    1   
1        Intel HD Graphics 6000  Other    1.34       898.94            0    0   

        cpu_name gpu_name  
0  Intel Core i5    Intel  
1  Intel Core i5    Intel

## Drop unnecessary columns

In [44]:
data = data.drop(columns=['laptop_ID', 'Inches', 'Product', 'ScreenResolution', 'Cpu', 'Gpu'])

In [45]:
data.head()

Company   TypeName  Ram  OpSys  Weight  Price_euros  Touchscreen  Ips  \
0   Apple  Ultrabook    8  Other    1.37      1339.69            0    1   
1   Apple  Ultrabook    8  Other    1.34       898.94            0    0   
2      HP   Notebook    8  Other    1.86       575.00            0    0   
3   Apple  Ultrabook   16  Other    1.83      2537.45            0    1   
4   Apple  Ultrabook    8  Other    1.37      1803.60            0    1   

        cpu_name gpu_name  
0  Intel Core i5    Intel  
1  Intel Core i5    Intel  
2  Intel Core i5    Intel  
3  Intel Core i7      AMD  
4  Intel Core i5    Intel

## One-Hot Encoding

In [46]:
data = pd.get_dummies(data)

In [47]:
data.head(2)

Ram  Weight  Price_euros  Touchscreen  Ips  Company_Acer  Company_Apple  \
0    8    1.37      1339.69            0    1             0              1   
1    8    1.34       898.94            0    0             0              1   

   Company_Asus  Company_Dell  Company_HP  ...  OpSys_Linux  OpSys_Other  \
0             0             0           0  ...            0            1   
1             0             0           0  ...            0            1   

   cpu_name_AMD  cpu_name_Intel Core i3  cpu_name_Intel Core i5  \
0             0                       0                       1   
1             0                       0                       1   

   cpu_name_Intel Core i7  cpu_name_Other  gpu_name_AMD  gpu_name_Intel  \
0                       0               0             0               1   
1                       0               0             0               1   

   gpu_name_Nvidia  
0                0  
1                0  

[2 rows x 30 columns]

In [48]:
data.shape

(1302, 30)

## Model Building and Selection

In [49]:
X = data.drop('Price_euros', axis=1)
y = data['Price_euros']

In [50]:
!pip install sklearn

  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.


## Divide dataset into training and testing set

In [52]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25)

In [53]:
X_train.shape, X_test.shape

((976, 29), (326, 29))

## model training

In [54]:
def model_acc(model):
    model.fit(X_train, y_train)
    acc = model.score(X_test, y_test)
    print(str(model)+ ' --> ' +str(acc))

In [55]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
model_acc(lr)

from sklearn.linear_model import Lasso
lasso = Lasso()
model_acc(lasso)

from sklearn.tree import DecisionTreeRegressor
dt = DecisionTreeRegressor()
model_acc(dt)

from sklearn.ensemble import RandomForestRegressor
rf = RandomForestRegressor()
model_acc(rf)

LinearRegression() --> 0.7111756572615779
Lasso() --> 0.7115710750787478
DecisionTreeRegressor() --> 0.7345044090288604
RandomForestRegressor() --> 0.7964526542468844


## Hyperparameter tunning

We are only going to use random forest model hereonwards

In [60]:
from sklearn.model_selection import GridSearchCV

parameters = {'n_estimators':[10, 50, 100],
              'criterion':['squared_error','absolute_error','poisson']}

grid_obj = GridSearchCV(estimator=rf, param_grid=parameters)

grid_fit = grid_obj.fit(X_train, y_train)

best_model = grid_fit.best_estimator_

best_model.score(X_test, y_test)

0.7982129932979167

In [61]:
X_test.columns

Index(['Ram', 'Weight', 'Touchscreen', 'Ips', 'Company_Acer', 'Company_Apple',
       'Company_Asus', 'Company_Dell', 'Company_HP', 'Company_Lenovo',
       'Company_MSI', 'Company_Other', 'Company_Toshiba',
       'TypeName_2 in 1 Convertible', 'TypeName_Gaming', 'TypeName_Netbook',
       'TypeName_Notebook', 'TypeName_Ultrabook', 'TypeName_Workstation',
       'OpSys_Linux', 'OpSys_Other', 'cpu_name_AMD', 'cpu_name_Intel Core i3',
       'cpu_name_Intel Core i5', 'cpu_name_Intel Core i7', 'cpu_name_Other',
       'gpu_name_AMD', 'gpu_name_Intel', 'gpu_name_Nvidia'],
      dtype='object')

## Save model

In [62]:
import pickle
with open('predictor.pickle', 'wb') as file:
    pickle.dump(best_model, file)